target/linux/ath79/dts/qca9557_dlink_dap-2660-a1.dts

```diff
@@ -0,0 +1,151 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "qca955x.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	model = "D-Link DAP-2660 A1";
+	compatible = "dlink,dap-2660-a1", "qca,qca9557";
+
+	aliases {
+		led-boot = &led_power_green;
+		led-failsafe = &led_power_red;
+		led-running = &led_power_green;
+		led-upgrade = &led_power_red;
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_power_red: power_red {
+			label = "red:power";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power_green: power_green {
+			label = "green:power";
+			gpios = <&gpio 23 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&spi {
+	status = "okay";
+
+	num-cs = <1>;
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <50000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "bootloader";
+				reg = <0x0 0x40000>;
+				read-only;
+			};
+
+			partition@40000 {
+				label = "bdcfg";
+				reg = <0x40000 0x10000>;
+				read-only;
+			};
+
+			partition@50000 {
+				label = "rgdb";
+				reg = <0x50000 0x10000>;
+				read-only;
+			};
+
+			partition@60000 {
+				label = "unknown";
+				reg = <0x60000 0x10000>;
+				read-only;
+			};
+
+			partition@70000 {
+				label = "firmware";
+				reg = <0x70000 0xee0000>;
+				compatible = "wrg";
+			};
+
+			partition@f50000 {
+				label = "langpack";
+				reg = <0xf50000 0x20000>;
+				read-only;
+			};
+
+			partition@f70000 {
+				label = "captival";
+				reg = <0xf70000 0x70000>;
+				read-only;
+			};
+
+			partition@fe0000 {
+				label = "certificate";
+				reg = <0xfe0000 0x10000>;
+				read-only;
+			};
+
+			art: partition@ff0000 {
+				label = "art";
+				reg = <0xff0000 0x10000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&uart {
+	status = "okay";
+};
+
+&mdio0 {
+	status = "okay";
+
+	#phy-mask = <0x1>;
+
+	phy0: ethernet-phy@4 {
+		reg = <0x4>;
+		phy-mode = "rgmii";
+	};
+};
+
+&eth0 {
+	status = "okay";
+
+	pll-data = <0x96000000 0x94000101 0x94001313>;
```

Comment:
So If speed 10 is unstable you need to try a bunch of different possible values for the cell for speed 10 which is the last one. Some possibilities replacing `0x94001313` with one of the following (other possibilities can be in other DTS files)

`0x94001616`
`0x00001313`
`0x00001616`
`0x08001313`
`0x08001616`
`0x0e001313`
`0x0e001616`
`0xa0001313`
`0xa0001616`

Reply:
We've actually tried _a lot_ of values for `pll-data` settings, either dumped out of the running oem bootloader (for individual link speeds), from DD-WRT or found in other .dts files, also using various delay settings for the `&gmac-config` register (at `0x18070000`).

These efforts eventually made 1000BASE-T work for DAP-2660<sup>[0]</sup> in the first place (unlike e.g. DAP-2360<sup>[1]</sup>, which uses ar7242 that does not even have a gmac-config register, but something yet different at `0x18050004`).
The CPU PLL is 720 MHz, but the clock does not seem to be so much of an issue here (for ar7242 I had even oscilloscoped the phy pins just to verify it's actually set to 125 / 25 MHz).

However, rather than randomly brute-forcing any values that happened to work for other devices,
I would appreciate if someone could point me towards a more deterministic approach on calculating these settings.

It's curious that neither the vendor nor DD-WRT seem to get all three link speeds working smoothly,
besides I wonder why there are three competing ways to setup internal delays, and which one(s) to choose:
 * on the phy itself, using `phy-mode = "rgmii-id"` / `"rgmii-rxid"` / `"rgmii-txid"` (with `rgmii-id` just setting both rx and tx internal delays)
 * in the `pll-data` value for the particular link speed (bit [31] for inversion, [29:28] for tx delay, [27:26] for rx delay)
 * in the `gmac-config` node (only present in newer devices), there is `txd-delay` (data from gmac to phy), `txen-delay` (tx enable, i.e. clock from gmac to phy), `rxd-delay` (data from phy to gmac) and `rxdv-delay` (data valid, i.e. clock signal from phy to gmac).

Is it just me lacking the skill to see through this, or is everyone else also just fiddling with settings until it works?
(what if none of these work and we'd need to implement initialisation differently in the kernel driver etc.)

I would thus be willing to donate one of each devices to someone having more experience with how these things work in the kernel drivers (preferably in Europe to avoid customs fees and delays).


[0] https://forum.openwrt.org/t/ath79-support-d-link-dap-2660-ethernet-not-working-in-gigabit-mode/71892
[1] https://forum.openwrt.org/t/ar8021-phy-not-working-in-gigabit-mode-how-to-configure-pll-data-cells/67207/2

Reply:
Yeah I'm one of the guys in those threads, since then I added 3 boards that need pll-data and also working on one that is not ready to add yet. I disagree that there exists a deterministic way to "find" the values you need, or at least, you need very fancy tools like oscilloscopes and a much more vast understanding of the SOC, which I don't have.

I achieved good performance on all speeds without any skills needed. Reading the registers in U-boot always had the wrong value for at least one of the speeds, and I found the rest with the guessing method. However I suppose you can read the datasheet and try to figure out which bits you need but that's a whole 32-bit word to decipher, at least half of them with a unique function.

for the 100 and 10 speeds there is always a pattern for the last bits: `0x101` for 100 and `0x1313` or `0x1616` for 10. At the very least you should try `0x94001616` if you haven't already. Good luck.